packages/flutter/lib/src/cupertino/tableview.dart

```diff
@@ -0,0 +1,485 @@
+import 'package:flutter/rendering.dart';
+
+@visibleForTesting
+const double kDefaultTableViewCellHeight = 44.0;
+
+@visibleForTesting
+const double kDefaultCellInset = 16.0;
+
+@visibleForTesting
+const double kDefaultDividerHeight = 1.0;
+
+/// WARNING: CupertinoTableView and its associated classes are EXPERIMENTAL. IF
+/// YOU USE THESE APIS, WE WILL BREAK YOU!
+///
+/// An iOS-styled table view.
+///
+/// Table views are used in iOS to render lists of cells. They are the iOS
+/// equivalent of Android's `ListView` and `RecyclerView`. [CupertinoTableView]
+/// is Flutter's iOS-styled alternative to [ListView].
+///
+/// #Cells
+///
+/// To achieve complete iOS style parity, consider filling [CupertinoTableView]s
+/// with [CupertinoTableViewCell]s. If custom cells are desired, any widget
+/// can be added to a [CupertinoTableView].
+///
+/// #Dividers
+///
+/// By default, [CupertinoTableView]s render dividers between each cell. For
+/// custom cells, the divider beneath the cell is rendered with a standard inset
+/// on the starting edge and is then drawn all the way to the ending edge. For
+/// [CupertinoTableViewCell]s, the divider starts where the title text starts
+/// and then renders all the way to the ending edge. This divider behavior is
+/// as per iOS behavior.
+///
+/// If a [CupertinoTableView] does not contain enough cells to fill the viewport
+/// then extra dividers are rendered beyond the last cell to the end of the
+/// viewport. This extra divider rendering extends into overscroll space at the
+/// end, too. The gap between extra dividers is equal to the height of the
+/// last cell in the [CupertinoTableView].
+///
+/// #Styles
+///
+/// A [CupertinoTableView] may be rendered either in a "plain" style or a
+/// "grouped" style.
+///
+/// A plain [CupertinoTableView] renders a single set of cells without any
+/// divisions between cells.
+///
+/// A grouped [CupertinoTableView] subdivides cells into sections. Sections
+/// are rendered with gaps between them and optional headers before each section.
+///
+/// #Section Index
+///
+/// If a [CupertinoTableView] is in a grouped style, it can optionally display
+/// a section "index". This index is a thin strip that appears on the right side
+/// of the [CupertinoTableView], displays titles for each section in the
+/// [CupertinoTableView], and allows users to tap on a given section title to
+/// automatically scroll to that section.
+///
+/// See also:
+///
+///  * [CupertinoTableViewCell], which renders an iOS-style table view cell that
+///    is intended to appear within a [CupertinoTableView].
+///  * [ListView], which renders a style agnostic list.
+class CupertinoTableView extends StatefulWidget {
+  /// Creates a plain-style [CupertinoTableView] with an explicit list of cells.
+  const CupertinoTableView.plain({
+    this.scrollController,
+    this.backgroundColor = CupertinoColors.white,
+    this.plainChildren = const <Widget>[],
+  }) : plainChildrenBuilder = null,
```

Comment:
I'd do something like the ListView here and create a delegate behind the constructor in the initializer. You won't be able to do const but I think it'll make the rest of the code simpler.

Reply:
A question for you and @Hixie - what is our comparative valuing of a const constructor vs a little cleaner implementation?  I think I came back and reworked this constructor specifically because I felt like it should be able to be const but it wasn't with my original implementation.  I can switch it back, but can you two confirm that the presence of a const constructor here is not a priority?

Reply:
Chatted offline. No const is fine. You're unlikely to invoke this constructor in a compile const way anyway. 

Ian also suggested to consider inheriting from BoxScrollView. This table does 2 things currently, draw divider and add extra trailing dividers at the end. 

The divider drawing wrapping can be done from the delegate and we'll have to create slivers for groups anyway. The trailing dividers can be done by a particular sliver instead of the root table itself. 

Anyway, doesn't have to be done in this PR but as you add more features, we might run into it naturally.